### ISSUE TITLE
No Proof Of Authentication

### ISSUE THREAD

**Issue by DaKnOb** [grnet/zeus#32]
Currently the voter has no proof that the website he is visiting is an unmodified version of this Open Source repository. That said, while the visuals remain the same, a malicious system administrator or attacker could replace the entire codebase with something different and then use a local copy of the software to generate all signatures and keys in his computer.

**Comment by louridas** [comment]
The trustee encryption keys are not stored in the server, so decryption would fail.

If the election officials have reason to believe that the administrator is malicious, they can always clone the repo and host the application themselves.

**Comment by DaKnOb** [comment]
So in order for this to work the system administrator must know the real trustees? Isn't there a way to trick them too into believing this process is authentic? Since he controls the software served to everyone and the software that generates the signatures (but not the one that verifies them), I believe that he can pull it off.
Sorry if I don't get the entire picture yet.

**Comment by louridas** [comment]
The trustees have their own keys. The system does not have their private keys (obviously). If we really want to follow the scenario, then the JS served to the trustees should also be malicious and leak the private keys to the administrator. Then again, the trustees can reuse keys that they have already generated, so the malicious administrator would have to work around that as well.

We could off course provide to the trustees a stand alone application that does nothing but generate keys, and have it digitally signed and made available to them. That's feasible and we may do that, which would require careful UI design, it should be cross-platform, etc.

Note though that experience has shown that the trustees are loth to any complication with the key generation process.

To put it another way. The real security danger, at least as we have seen it to now, is that the trustees do not really pay attention to what they do with their keys. We are sure you know that you can add security layers, with the risk that the inconvenience will raise the risk of human error or inaction. So the whole UX and usability side of things plays a larger role to what we roll out than it might seem.

**Comment by DaKnOb** [comment]
If there was a way to easily verify that the code running is the one from this repository then the problem would be solved, but it's okay. It's usability over security. Maybe this one is a bit paranoid.

**Comment by louridas** [comment]
In general, verifying that a VM runs a specified piece of software is an open research problem. 

In our particular case, and concerning the client side JS code only, we could try an approach like [Mylar](https://css.csail.mit.edu/mylar/), ported to Python somehow (as the original is over Node).
